File: app/src/main/java/es/etologic/mahjongscoring2/app/game/activity/GameActivityViewModelFactory.java
package es.etologic.mahjongscoring2.app.game.activity;

import android.arch.lifecycle.ViewModel;
import android.arch.lifecycle.ViewModelProvider;
import android.support.annotation.NonNull;

import javax.inject.Inject;

import es.etologic.mahjongscoring2.domain.use_cases.CreateGameUseCase;
import es.etologic.mahjongscoring2.domain.use_cases.GetGamesUseCase;
import es.etologic.mahjongscoring2.domain.use_cases.UpdateRoundsUseCase;

public class GameActivityViewModelFactory extends ViewModelProvider.NewInstanceFactory {

    private CreateGameUseCase createGameUseCase;
    private final GetGamesUseCase getGamesUseCase;
    private UpdateRoundsUseCase updateRoundUseCase;


    @Inject
    GameActivityViewModelFactory(CreateGameUseCase createGameUseCase, GetGamesUseCase getGamesUseCase, UpdateRoundsUseCase updateRoundUseCase) {
        this.createGameUseCase = createGameUseCase;
        this.getGamesUseCase = getGamesUseCase;
        this.updateRoundUseCase = updateRoundUseCase;
    }

    @SuppressWarnings("unchecked")
    @NonNull
    @Override
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        return (T) new GameActivityViewModel(createGameUseCase, getGamesUseCase, updateRoundUseCase);
    }
}

File: app/src/main/java/es/etologic/mahjongscoring2/app/game/activity/GameActivityViewModelFactory.kt
package es.etologic.mahjongscoring2.app.game.activity

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import es.etologic.mahjongscoring2.domain.use_cases.EndGameUseCase
import es.etologic.mahjongscoring2.domain.use_cases.GetCurrentGameUseCase
import es.etologic.mahjongscoring2.domain.use_cases.UpdateGameUseCase
import javax.inject.Inject

class GameActivityViewModelFactory @Inject internal constructor(
    private val getCurrentGameUseCase: GetCurrentGameUseCase,
    private val updateGameUseCase: UpdateGameUseCase,
    private val endGameUseCase: EndGameUseCase
) : ViewModelProvider.NewInstanceFactory() {
    
    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        return GameActivityViewModel(getCurrentGameUseCase, updateGameUseCase, endGameUseCase) as T
    }
}
